File: PAT_Basic_Level_by_lua/1059 C语言竞赛/1059 C语言竞赛.lua
-- 题目编号: 1059 C语言竞赛
-- 实现原理: 用哈希表存储参赛者ID到排名的映射。定义素数判断函数。对每个查询，若ID不在哈希表中则输出"Are you kidding?"；若已查询过则输出"Checked"；否则根据排名输出对应奖项：第1名"Mystery Award"，素数排名"Minion"，其他"Chocolate"。
-- n: 参赛人数，r: ID到排名的映射表
local n, r = tonumber(io.read()), {}
-- 读取参赛者ID，按顺序记录排名（第i名）
for i = 1, n do
    r[io.read("l")] = i
end
-- 素数判断函数
local function p(x)
    if x < 2 then return false end  -- 小于2不是素数
    -- 从2到sqrt(x)遍历，若能整除则不是素数
    for i = 2, math.floor(math.sqrt(x)) do
        if x % i == 0 then return false end
    end
    return true
end
-- seen: 记录已查询过的ID，避免重复颁奖
local seen = {}
-- 处理查询
for _ = 1, tonumber(io.read()) do
    local x = io.read("l")  -- 查询的ID
    if not r[x] then
        -- ID不存在
        print(x .. ": Are you kidding?")
    elseif seen[x] then
        -- 已查询过
        print(x .. ": Checked")
    else
        seen[x] = true  -- 标记为已查询
        -- 根据排名颁奖：第1名Mystery Award，素数排名Minion，其他Chocolate
        print(x .. ": " .. (r[x] == 1 and "Mystery Award" or p(r[x]) and "Minion" or "Chocolate"))
    end
end

File: PAT_Basic_Level_by_perl/1059 C语言竞赛/1059 C语言竞赛.pl
use strict;
use warnings;

# =====================================================================
# 题目：1059 C语言竞赛
# 难度：中等
# 知识点：素数筛, 哈希表
# =====================================================================
#
# 【题目描述】
#   给定比赛最终排名（按顺序给出一系列参赛者 ID）和若干需要查询的 ID。
#   冠军得 Mystery Award，排名为素数的得 Minion，其余得 Chocolate；
#   不在排名中输出 Are you kidding?，已查询过的输出 Checked。
#
# 【实现原理】
#   1. 用哈希表记录每个 ID 对应的排名（1 为冠军）。
#   2. 用埃氏筛预处理 1..N 的素数表。
#   3. 依次处理查询：先判是否在排名中，再判是否查过，最后按排名给奖。
#
# 【输入样例】
#   6
#   1111
#   6666
#   8888
#   1234
#   5555
#   0001
#   6
#   8888
#   0001
#   1111
#   2222
#   8888
#   2222
# 【输出样例】
#   8888: Minion
#   0001: Chocolate
#   1111: Mystery Award
#   2222: Are you kidding?
#   8888: Checked
#   2222: Are you kidding?
# =====================================================================

my @tok = split /\s+/, join('', <STDIN>);   # 读取全部 token（保留前导零）
my $idx = 0;
my $n = $tok[$idx++];

# 记录每个 ID 的排名（1 为冠军）
my %rank;
for my $r (1 .. $n) {
    $rank{$tok[$idx++]} = $r;
}

my $k = $tok[$idx++];
my @queries = @tok[$idx .. $idx + $k - 1];

# 埃氏筛求 1..N 的素数
my @prime = (1) x ($n + 1);
$prime[0] = $prime[1] = 0;
for my $i (2 .. int(sqrt($n))) {
    next unless $prime[$i];
    for (my $j = $i * $i; $j <= $n; $j += $i) {
        $prime[$j] = 0;
    }
}

my %seen;
for my $id (@queries) {
    if (!exists $rank{$id}) {
        print "$id: Are you kidding?\n";
    } elsif ($seen{$id}) {
        print "$id: Checked\n";
    } else {
        my $r = $rank{$id};
        my $prize = $r == 1 ? "Mystery Award"
                  : $prime[$r] ? "Minion"
                  : "Chocolate";
        print "$id: $prize\n";
        $seen{$id} = 1;
    }
}
